src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateCopy = { ...state };
+  const stateHistory = [];
+
+  for (let index = 0; index < actions.length; index++) {
+    const action = actions[index];
+    const type = action.type;
+    const keys = Object.keys(stateCopy);
+
+    switch (type) {
+      case 'addProperties':
+        stateCopy = { ...stateCopy, ...action.extraData };
+        stateHistory.push({ ...stateCopy });
+        break;
+
+      case 'removeProperties':
+        for (let i = 0; i < action.keysToRemove.length; i++) {
+          delete stateCopy[action.keysToRemove[i]];
+        }
+        stateHistory.push({ ...stateCopy });
+        break;
+
+      case 'clear':
+        for (let i = 0; i < keys.length; i++) {
+          delete stateCopy[keys[i]];
+        }
+        stateHistory.push({});
+    }
```

Comment:
[CODE STYLE] Use a `default` case in your `switch` statement for error handling to ensure that there is a response for unhandled `type` values.